topsites.py

```diff
@@ -0,0 +1,166 @@
+}
+
+# Cache parsed sites, change priority if raised
+topsites = {}
+
+engine = create_engine('sqlite:///topsites.db', echo=True)
+Base = declarative_base()
+Session = sessionmaker(bind=engine)
+session = Session()
+
+
+# sqlalchemy base object for a top site
+class Site(Base):
+    __tablename__ = "topsites"
+
+    url = Column(String, primary_key=True)
+    priority = Column(Integer)
+    country_code = Column(String)
+    ranking = Column(Integer)
+
+    def __init__(self, url, priority, country_code, ranking):
+        self.url = url
+        self.priority = priority
+        self.country_code = country_code
+        self.ranking = ranking
+
+    def __repr__(self):
+        return "<Site('%d','%d' '%s', '%s')>" % (self.priority, self.ranking, self.url, self.country_code)
+
+Base.metadata.create_all(engine)
+
+
+# TopSites is used to grab 100 sites from alexa top site API
+class TopSites:
+    def __init__(self, access_key, secret_key):
+        self.access_key = access_key
+        self.secret_key = secret_key
+
+    def query(self, begin, begin_priority):
+        query_string = self.build_query(begin, 100)
+        toSign = "GET\n" + ATS_SERVICE_HOST + "\n/\n" + query_string
+        signature = self.gen_sign(toSign, self.secret_key)
+        uri = ATS_AWS_BASE_URL + query_string + "&" + urlencode({"Signature": signature}, "UTF-8")
+
+        response = requests.get(uri)
+
+        if response.status_code == requests.codes.ok:
+            # See http://docs.aws.amazon.com/AlexaTopSites/latest/index.html?QUERY_QueryRequests.html
+            dom = parseString(response.content)
+            sites = dom.getElementsByTagName("aws:Site")
+            for site in sites:
+                url = self.get_text(site.getElementsByTagName("aws:DataUrl")[0].childNodes)
+                rank = int(self.get_text(site.getElementsByTagName("aws:Rank")[0].childNodes))
+
+                priority = begin_priority
+                if rank > 100 and rank <= 1000:
+                    priority = priority + 1
+                elif rank > 1000 and rank <= 10000:
+                    priority = priority + 2
+
+                if url in topsites:
+                    site_row = topsites[url]
+                    # Update priority is new is lower
+                    if site_row.priority > priority:
+                        site_row.priority = priority
+                        site_row.country_code = self.country_code
+                        site_row.ranking = rank
+                else:
+                    site_row = Site(url, priority, self.country_code, rank)
+                    topsites[url] = site_row
+                    session.add(site_row)
+
+            session.commit()
+
+    # Build query string for alexa top site
+    # with start ranking and count
+    def build_query(self, start, count):
+        now = time.time()
+        mlsec = repr(now).split('.')[1][:3]
+        timestamp = datetime.datetime.utcnow().strftime(ATS_DATEFORMAT).format(mlsec)
```

Comment:
style nit: line too long.

```python
        timestamp = datetime.datetime.utcnow().strftime(
            ATS_DATEFORMAT).format(mlsec)
```

Though I would probably separate in two lines myself. Too many chained functions are not very pleasant on the eye.

```python
        now_date = datetime.datetime.utcnow()
        timestamp = now_date.strftime(ATS_DATEFORMAT).format(mlsec)
```